File: 496-next-greater-element-i/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int,int> next;
        stack<int> st;
        vector<int> ans;

        for (int i=nums2.size()-1;i>=0;i--){
            int num=nums2[i];
            while(!st.empty() && st.top()<=num){
                st.pop();

            }
            if ( !st.empty()){
                next[num]=st.top();

            }
            else{
                next[num]=-1;

            }
            st.push(num);

        }
        for(int num : nums1){
            ans.push_back(next[num]);

        }

        return ans;
    }
};

File: 496-next-greater-element-i/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        st=[]
        ans=[]
        nextt={}
        for  i in range(len(nums2)-1,-1,-1):
            num=nums2[i]
            #st.append(num)
            while st and st[-1]<=num:
                st.pop()
            if st :
                nextt[num]=st[-1]
            else:
                nextt[num]=-1
            st.append(num)    
                
        for i in nums1:
            ans.append(nextt[i])
        return ans                    